content/docs/reference/design-docs/editoast-errors/_index.en.md

```diff
@@ -0,0 +1,320 @@
+
+````rust
+fn get_resource(id: u64) -> Result<Resource, Enum3<DbError, RedisError, MissingResourceError>> {
+    todo!()
+}
+
+fn process_resource(resource: Resource) -> Result<Computation, Enum2<DbError, ProcessingError> {
+    todo!()
+}
+
+#[utoipa::path(
+    ...,
+    responses(
+        (status = 200, body = Computation),
+        (status = 400, body = EndpointError)
+    )
+)]
+async fn endpoint(Path(key): Path<Key>) -> Result<Json<Computation>, EndpointError> {
+    let resource = get_resource(key)?;
+    let value = process_resource(resource)?;
+    Ok(Json(value))
+}
+
+#[derive(Debug, thiserror::Error, EditoastError)]
+pub enum EndpointError {
+    Db(#[from] DbError),
+
+    Redis(#[from] RedisError),
+
+    #[error(transparent)]
+    #[editoast_error(user)]
+    Missing(#[from] MissingResourceError)
+
+    #[error(transparent)]
+    #[editoast_error(user)]
+    Processing(#[from] ProcessingError)
+}
+````
+
+The implementation of the `EnumX` type would be rather easy to generate for many tuple sizes. The crate [anon_enum](https://docs.rs/anon_enum/1.1.0/anon_enum/) exists but if we choose to use this pattern, it's probably better to have our own type for greater flexibility and avoid another dependency.
+
+## Full `derive(EditoastError)` spec
+
+The macro supports enums, named structs and tuple structs (fixme correct naming).
+
+````rust
+#[derive(thiserror::Error)] // not an EditoastError
+#[error("wrong string: {0}")]
+struct WrongString(String);
+
+#[derive(EditoastError, thiserror::Error)]
+#[error("wrong int: {value}")]
+#[editoast_error(user, name = "InvalidInt")]
+struct WrongInt { value: i64 };
+
+#[derive(EditoastError, thiserror::Error)]
+#[error("my error type")]
+#[editoast_error(context)]
+enum MyError {
+    // error_type = "MyError::InvalidString"
+    // context = { expected_format: string }
+    // #[editoast_error(internal)] by default
+    InvalidString { #[from] source: WrongString, expected_format: String },
+
+    // error_type = "MyError::InvalidInt"
+    // context = {} (xyz is skipped as we forward the editoast error)
+    WrongInt { #[from] #[editoast_error] source: WrongInt, xyz: String }
+
+    // error_type = "MyError::Bad"
+    // context = { "0": string, "1": number }
+    #[error("user did a bad with {0} and {1}")]
+    #[editoast_error(user, name = "Bad")]
+    Oops(String, i64)
+}
+````
+
+### Providing `context`
+
+Only `4xx` errors require context since these are the errors the frontend may use to provide the user some form of recovery. Context is computed just before the error is serialized in `axum`'s error handler.
+
+`derive(EditoastError)` provides a few ways to set it.
+
+````rust
+#[derive(Debug, thiserror::Error, EditoastError)]
+enum Error {
+    #[editoast_error(user, context)]
+    No { because: String }, // { }
```

Comment:
Not clear to me why no context is produced here but the `Nein` variant which is basically the same (a `Vec` instead of a `String`) does have a context. Is it just because the name is `reasons`?

Reply:
Good point, the annotation should have been `#[editoast_error(user)]`.